File: #2. 2019 카카오 개발자 겨울 인턴십/김민정/problem2_튜플.java
// 풀이 시간: 이것도 기억이 잘 안 나지만,,, 1번보다는 확실히 오래걸림..
// 시간 복잡도: O(L + N log N) (L : s의 길이, N : 숫자의 종류)
// 코드 설명: 입력받은 문자열을 돌면서 각 숫자가 몇 번 나왔는지 셈 -> 가장 많이 등장한 숫자부터 리턴배열에 넣음
// 취약한 부분: Map에 들어있는 값들을 value 기준으로 정렬해야 했는데 이 과정이 처음이라 어려움이 있었음

import java.util.*;

class Solution {
    public int[] solution(String s) {

        Map<Integer, Integer> numCnt = new HashMap<>();
        StringBuilder num = new StringBuilder();
        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            if (ch-'0' >= 0 && ch-'0' <= 9) {
                num.append(ch);
            } else {
                if (num.length() != 0) {
                    int tmp = Integer.parseInt(num.toString());
                    numCnt.put(tmp, numCnt.getOrDefault(tmp, 0) + 1);
                    num.setLength(0);
                }
            }
        }
        
        List<Map.Entry<Integer, Integer>> list = new ArrayList<>(numCnt.entrySet());
        list.sort(Collections.reverseOrder(Map.Entry.comparingByValue()));
        
        int[] answer = new int[numCnt.size()];
        int idx = 0;
        for (Map.Entry<Integer, Integer> entry : list) {
            answer[idx++] = entry.getKey();
        }        
        
        return answer;
    }
}

File: #2. 2019 카카오 개발자 겨울 인턴십/장효빈/problem2_튜플.py
# 풀이 시간: 41분
# 시간 복잡도: O(s)
# 코드 설명: 입력을 쪼개서 숫자로 바꾸고 길이순 정렬 후 원소를 하나씩 체크해 set에 없으면 정답 배열과 set에 추가
# 취약한 부분: 정규표현식을 써보려고 했는데 파이썬이라 익숙하지가 않아서 시간만 날렸다

def solution(s):
    l = []
    s = s[2:-2].split(r"},{")
    for e in s:
        l.append(list(map(int, e.split(','))))
    
    l.sort(key = lambda x: len(x))
    
    elements = set()
    answer = []
    for e in l:
        for element in e:
            if element not in elements:
                answer.append(element)
                elements.add(element)
            
    return answer